### PR TITLE
Have shadow plugin relocate source references too by defining a separ…

### PR DESCRIPTION
**Pull request by anuraaga**
…ate task for it. Only run shading and proguarding when building a jar.

When doing some recent work, I found it annoying that

- I had no sources for relocated dependencies I wanted to use and understand
- Running tests from command line always ran a very slow proguard task

The main reason for this seemed to be us only running shadow on dependencies, not on the armeria source itself. This is probably because we didn't want shadow to bundle all the dependencies as it usually does. Indeed, I couldn't find a good way to only bundle the relocated deps along with proguard - it was either all or nothing.

I found this approach that adds an additional shadow step that runs on the source with no bundling and overwrites the jar contents with it. It appears to be working. I have done the rewriting in all packages since currently it seems we have a style of having internal dependencies in core, and having other packages like thrift use them. If we ever had subpackage-specific internal dependencies, there'd need to be similar tasks added to the subpackage but otherwise should work ok.

Please patch this in and make sure it's doing what you expect the shading to do. Any suggestions on making it cleaner / more future-proof are of course welcome.

Result: Easier development since weirdness is all moved to the packaging step.

Extra change: I moved the trimmed internal dependencies into the core build directory. I wasn't sure why you put them in the root project, with the added work of extending the ```clean``` task. Let me know if I should revert this too.

### CODE REVIEW

**Comment by anuraaga** [comment]
PTAL @trustin (looks like with read access, you can assign reviews to me, but I can't assign to you. Oh well :) )

File: core/build.gradle [hunk 1]
```diff
@@ -55,22 +56,11 @@ dependencies {
...
-
-clean {
-    delete internalDependenciesJarTrimmed
-    delete internalDependenciesJarRelocated
 }
 
 task trimInternalDependencies(type: ProGuardTask) {
     injars files(configurations.internalDependency.collect()), filter: '!META-INF/**'
     outjars internalDependenciesJarTrimmed
-    libraryjars files(configurations.runtime.collect())
```
Comment (anuraaga) [inline core/build.gradle]: Ah accidentally left this change in - I made it while experimenting with a different approach. But found the result seems to be the same without it. What do you think?

Comment (trustin) [inline core/build.gradle]: In my case, ProGuard was trimming everything away without it, but I may be wrong.

**Comment by trustin** [comment]
> Extra change: I moved the trimmed internal dependencies into the core build directory. I wasn't sure why you put them in the root project, with the added work of extending the clean task. Let me know if I should revert this too.

SGTM. No idea why I did that. Maybe fixing build script was poisonous to me? :-D

**Comment by anuraaga** [comment]
Looking at much bytecode, think I found the issue. It seems that our current proguard configuration isn't ideal.

The proguard configuration doesn't preserve the Int2ObjectFunction interface, which is where ```put(int, Object)``` is defined. So all we're left with is the normal method from Map ```put(Object, Object)```. When this is called from code, the int parameter is boxed.

This means proguard is "unfasting" fastutil as avoiding boxing is one of the main reasons to use it. The reason this doesn't cause a compile error in master is we compile against the proguarded classes, while in this PR it's changed so our source is compiled against the real classes and then references are rewritten to the proguarded ones.

I added keepinterfaces to preserve the interfaces we need. I wish proguard had an option, keep all implemented interfaces, but I couldn't fine one.

**Comment by trustin** [comment]
One problem I'm concerned of is the test will never run with the shaded version, which means there's no way to tell if our shading process went well or not. Should we run the tests twice when building a shaded JAR?

**Comment by anuraaga** [comment]
Yeah it's a valid concern - what I found is that

```./gradlew :core:test```

which is often used in development won't shade but

```./gradlew :core:build```

will first build the jar and then run the tests, which means they use the shaded version created during jar generation. build is used from the CI.

Does this seem like a reasonable result?

**Comment by trustin** [comment]
> Does this seem like a reasonable result?

Ha! That's a nice side effect. SGTM.

**Pull request merged**